twilight-gateway/src/future.rs

```diff
@@ -36,44 +40,56 @@ pub enum NextMessageFutureOutput {
     UserChannelMessage(Message),
 }
 
-/// Future to determine the next action when [`Shard::next_message`] is called.
-///
-/// Polled futures are given a consistent precedence, from first to last polled:
-///
-/// - [sending a heartbeat to Discord][1];
-/// - [relaying a user's message][2] over the Websocket message;
-/// - [receiving a message][3] from Discord
-///
-/// **Be sure** to keep documented precedence in sync with variants in
-/// [`NextMessageFutureOutput`]!
-///
-/// [1]: NextMessageFutureOutput::SendHeartbeat
-/// [2]: NextMessageFutureOutput::UserChannelMessage
-/// [3]: NextMessageFutureOutput::Message
-/// [`Shard::next_message`]: crate::Shard::next_message
-pub struct NextMessageFuture<'a> {
-    /// Future resolving when the user has sent a message over the channel, to
-    /// be relayed over the Websocket connection.
-    channel_receive_future: &'a mut UnboundedReceiver<Message>,
-    /// Future resolving when the next Websocket message has been received.
-    message_future: Next<'a, Connection>,
-    /// Future resolving when the [`Shard`] must sent a heartbeat.
+pin_project! {
+    /// Future to determine the next action when [`Shard::next_message`] is
+    /// called.
+    ///
+    /// Polled futures are given a consistent precedence, from first to last:
     ///
-    /// [`Shard`]: crate::Shard
-    tick_heartbeat_future: TickHeartbeatFuture,
+    /// - [sending a heartbeat to Discord][1];
+    /// - [relaying a user's message][2] over the Websocket message;
+    /// - [receiving a message][3] from Discord
+    ///
+    /// **Be sure** to keep documented precedence in sync with variants in
+    /// [`NextMessageFutureOutput`]!
+    ///
+    /// [1]: NextMessageFutureOutput::SendHeartbeat
+    /// [2]: NextMessageFutureOutput::UserChannelMessage
+    /// [3]: NextMessageFutureOutput::Message
+    /// [`Shard::next_message`]: crate::Shard::next_message
+    pub struct NextMessageFuture<'a, F, St> {
+        // Future resolving when the user has sent a message over the channel,
+        // to be relayed over the Websocket connection.
+        channel_receive_future: &'a mut UnboundedReceiver<Message>,
+        // Future resolving when the next Websocket message has been received.
+        message_future: Next<'a, St>,
+        // Future resolving to a ratelimit permit, if a ratelimiter is enabled.
+        #[pin]
+        ratelimit_permit: Option<F>,
```

Comment:
Could potentially avoid this by stack pinning the future in `Shard::next_message`, but I couldn't make it work